Remove loadConfiguration where it is possible

palagdan opened the pull request:
Resolves #274. 

I removed `loadConfiguration()` where possible in `s-pipes-modules`. However, there are three main reasons why I couldn't remove it from some other modules:

- 1) Field dependency on runtime values: Some fields depend on values retrieved from previous fields, which is why these values can only be defined at runtime.
 ```java
  @Override
    public void loadConfiguration() {
        e5xResourceUriStr = getEffectiveValue(KBSS_MODULE.JENA.has_resource_uri).asLiteral().toString();
        e5xResource = getResourceByUri(e5xResourceUriStr);
    }
```

- 2) Dependency on parent module configuration: Some modules invoke loadConfiguration() from parent modules.
```java
   @Override
    public void loadConfiguration() {
        super.loadConfiguration();
        marginalConstraint = getEffectiveStringValue(TYPE_PREFIX + "marginal-constraint");
        marginalsDefsFileUrl = getEffectiveStringValue(TYPE_PREFIX + "marginals-defs-file-url");
        marginalsFileUrl = getEffectiveStringValue(TYPE_PREFIX + "marginals-file-url");
        dataServiceUrl = getEffectiveStringValue(TYPE_PREFIX + "data-service-url");
    }
```
- 3) Conditional field initialization: Some fields with parameters need to be initialized only under certain conditions.
```java
    @Override
    public void loadConfiguration() {

        if (this.resource.getProperty(DescriptorModel.JENA.has_document_date) != null) { // TODO set current date if not specified
            documentDate = getEffectiveValue(DescriptorModel.JENA.has_document_date).asLiteral().toString();
        }

        if (this.resource.getProperty(DescriptorModel.JENA.has_rule_file) != null) { //TODO support more rule files
            ruleFilePaths.add(Paths.get(getEffectiveValue(DescriptorModel.JENA.has_rule_file).asLiteral().toString()));
        }
    }
```

blcham commented:
- It seems like i) and iii) are not are not issue.
- iii) should be solved by calling function to load configuration after @Parameter annotations are processed

palagdan commented:
@blcham

I have removed all loadConfiguration() methods from the modules and replaced them with loadManualConfiguration(). Also I removed manual configuration for fields that can be configured using annotations.

blcham commented on s-pipes-modules/module-nlp/src/main/java/cz/cvut/spipes/modules/SUTimeModuleNew.java:
```diff
@@ -72,30 +78,6 @@ public String getTypeURI() {
         return TYPE_URI;
     }
 
-    public int getPageSize() {
```
as mentioned before getter/setter mismatch with:
`private Integer pageSize = DEFAULT_PAGE_SIZE;`

palagdan replied:
done

blcham commented on s-pipes-modules/module-sparql-endpoint/src/main/java/cz/cvut/spipes/modules/DownloadGraphModule.java:
```diff
@@ -39,50 +44,10 @@ public class DownloadGraphModule extends AnnotatedAbstractModule {
-        this.endpointUrl = endpointUrl;
-    }
-
-    public int getPageSize() {
```
Mismatch with `private Integer pageSize = DEFAULT_PAGE_SIZE;`

palagdan replied:
done

palagdan replied:
done

blcham commented on s-pipes-modules/module-tarql/src/main/java/cz/cvut/spipes/modules/TarqlModule.java:
```diff
@@ -142,26 +142,22 @@ public String getTypeURI() {
-
-
-
+//    @Override
```
should not be commented out, moreover there is log.debug

palagdan replied:
t’s fully configured via annotations. I chose to comment it out rather than remove it because there were TODO comments related to this configuration. I can leave then:
```java
  @Override
    public void loadManualConfiguration() {
        log.debug("Loaded {} spin construct queries.", constructQueries.size());
    }
```

palagdan replied:
done

blcham commented on s-pipes-modules/module-text-analysis/src/main/java/cz/cvut/spipes/modules/TextAnalysisModule.java:
```diff
@@ -187,9 +187,8 @@ private String annotateObjectLiteral(String objectValue) {
     @Override
-    public void loadConfiguration() {
+    public void loadManualConfiguration() {
         super.loadConfiguration();
```
!

palagdan replied:
I don't quite understand why there is a super.loadConfiguration() line. The code should work the same without loadConfiguration() or loadConfigurationManual() since TextAnalysisModule extends AnnotatedAbstractModule. Therefore, we can remove loadManualConfiguration().

palagdan replied:
done

palagdan commented on s-pipes-modules/module-sparql-endpoint/src/main/java/cz/cvut/spipes/modules/ApplyConstructAbstractModule.java:
```diff
@@ -167,23 +137,23 @@ protected String substituteQueryMarkers(int currentIteration, String queryStr) {
-        parseText = this.getPropertyValue(KBSS_MODULE.JENA.is_parse_text, true);
-        iterationCount = this.getPropertyValue(KBSS_MODULE.JENA.has_max_iteration_count, -1);
-    }
+//    @Override
```
Why should it be enabled, if it is totally configured by annotations?

palagdan replied:
done

palagdan commented on s-pipes-modules/module-sparql-endpoint/src/main/java/cz/cvut/spipes/modules/ApplyConstructWithChunkedValuesAndScrollableCursorModule.java:
```diff
@@ -106,13 +106,9 @@ protected String substituteQueryMarkers(int currentIteration, String queryStr) {
     @Override
-    public void loadConfiguration() {
+    public void loadManualConfiguration() {
         super.loadConfiguration();
```
I will check other places. The motivation behind this is that the parent of ApplyConstructWithChunkedValuesAndScrollableCursorModule is ApplyConstructAbstractModule, which is fully configured by loadConfiguration()

palagdan replied:
Done, all modules don't use l`oadConfiguration()`. I think loadManualConfiguration makes more sense to me because it indicates that you should do it yourself.

palagdan commented on s-pipes-modules/module-sparql-endpoint/src/main/java/cz/cvut/spipes/modules/RetrieveGraphModule.java:
```diff
@@ -29,35 +33,10 @@ public class RetrieveGraphModule extends AnnotatedAbstractModule {
-        this.endpointUrl = endpointUrl;
-    }
-
-    public int getPageSize() {
```
Lombok will generate setter and getter for pageSize. That is why I removed boilerplate code from this module.

palagdan replied:
done

palagdan commented on s-pipes-modules/module-rdf4j/src/main/java/cz/cvut/spipes/modules/Rdf4jDeployModule.java:
```diff
@@ -67,7 +67,7 @@ private static Property getParameter(final String name) {
         comment = "IRI of contexts is inferred from annotated input triples. Only reified triples that contain triple " +
             "?reifiedStatement kbss-module:is-part-of-graph ?graph are processed." +
             " Actual triples related to reified statement are not processed/needed.")
```
it will be done!

palagdan replied:
done

palagdan commented:
@blcham 
I believe we can easily replace Integer with int because it is only used in one place, as follows:
```
module.setPageSize(10);
```

palagdan commented:
@blcham 

I’m not sure why, but when I try to test locally with `mvn clean install`, I get the same error that we solved in a previous ticket. However, all the tests pass in the GitHub environment. Do you have any suggestions on how I can fix this?

![error_parameter](https://github.com/user-attachments/assets/3d379dde-07ff-4b56-8eac-f7101835ef85)

palagdan commented:
@blcham 

Also I stucked on the problem with Dependency on parent module configuration.

there is 
```java
class ApplyConstructAbstractModule extends AnnotatedAbstractModule 
```

Some modules extend this class to have configured fields, such as constructQueries, isReplace, and parseText.

for example this module:
```java
class ApplyConstructWithChunkedValuesModule extends ApplyConstructAbstractModule
```

As it was before, the ApplyConstructWithChunkedValuesModule called super.loadConfiguration(), which was overridden in ApplyConstructAbstractModule to manually configure the fields.

However, when I remove this overriding method from the parent, I’m unsure how to configure it from the child class using annotations. If we call super.loadConfiguration() from the child, it will see that the parent did not implement it and will then call loadConfiguration from AnnotatedAbstractModule, which only configures the child class.

Therefore, I’m not sure how to resolve the problem without manually configuring all the fields in ApplyConstructAbstractModule within the loadManualConfiguration() method and then calling it from the child modules.

blcham commented:
> I’m not sure why, but when I try to test locally with mvn clean install, I get the same error that we solved in a previous ticket.

Maybe you have local changes there, try this to be sure:
```
git clone -b 274-load-configuration-parallel 'https://github.com/kbss-cvut/s-pipes'
cd s-pipes
mvn clean install
```

-----

palagdan commented:
@blcham 

Thank you for the idea! It should be better than doing it manually.

palagdan commented:
> But how about reimplementing loadConfiguration() to process also parent classes? Would it solve the issue?
> I mean to process @parameter annotations of all classes from something like `this.getClass().getSuperclass()`.

@blcham 

Thank you for the idea! It should be better than doing it manually.

blcham commented on s-pipes-core/src/main/java/cz/cvut/spipes/modules/AnnotatedAbstractModule.java:
```diff
@@ -18,40 +18,86 @@
+     * {@link #loadManualConfiguration()} method to allow subclasses to handle any custom configuration
+     * that requires manual intervention.
+     *
+     * @throws RuntimeException if two or more fields in the class hierarchy share the same
```
See my note above.

palagdan replied:
done

blcham reviewed (changes requested):
Almost done, see my comments

blcham commented on s-pipes-core/src/main/java/cz/cvut/spipes/modules/AnnotatedAbstractModule.java:
```diff
@@ -18,40 +18,86 @@
+     *
+     * <p>For each annotated field:
+     * <ul>
+     *   <li>A {@link Setter} is selected based on the field's type (e.g., {@link ListSetter} for lists).</li>
```
So person understand what is `field's type`

palagdan replied:
done

palagdan commented on s-pipes-core/src/main/java/cz/cvut/spipes/modules/AnnotatedAbstractModule.java:
```diff
@@ -18,40 +18,86 @@
+     * <p>This method is responsible for processing the fields declared in the class and its superclasses,
+     * that are annotated with {@link Parameter}. It ensures that each field is processed only once and
+     * verifies that no two fields share the same {@link Parameter#iri()}. If a duplicate is found,
+     * it throws a {@link RuntimeException}.
```
done

palagdan commented on s-pipes-core/src/main/java/cz/cvut/spipes/modules/AnnotatedAbstractModule.java:
```diff
@@ -18,40 +18,86 @@
+     *       which is responsible for setting the value of the field based on its {@link Parameter#iri()}.</li>
+     * </ul>
+     *
+     * <p>The method continues processing all declared fields, moving up through the class hierarchy
```
I was thinking about it, when I implemented it.  I re implemented so it traverses downwards now.

blcham replied:
Do not like that much explicit stack, but It is ok.

Instead of this, I would rather use recursion.
```
private void processClassHierarchy(Class<? extends AnnotatedAbstractModule> clazz) {
    if (clazz == AnnotatedAbstractModule.class) {
        return;
    }
    
    // Recur to the superclass first
    processClassHierarchy((Class<? extends AnnotatedAbstractModule>) clazz.getSuperclass());
    
    // Now process the current class
    // e.g. some logic involving 'clazz'
}
```

blcham reviewed (approved):
LGTM, you can merge it without change, just see my comment about recursion

blcham commented:
@palagdan just for next time:

![image](https://github.com/user-attachments/assets/02e40fa6-8dab-4100-85fe-f81ceb23754d)
